src/lit-html.ts

```diff
@@ -12,21 +12,39 @@
  * http://polymer.github.io/PATENTS.txt
  */
 
+/**
+ * TypeScript has a problem with precompiling templates literals
+ * https://github.com/Microsoft/TypeScript/issues/17956
+ */
+const shouldCacheTemplates = ((_t: any) => _t() === _t())(() => ((s: any) => s)``);
+
 // The first argument to JS template tags retain identity across multiple
 // calls to a tag for the same literal, so we can cache work done per literal
 // in a Map.
 const templates = new Map<TemplateStringsArray, Template>();
+const cacheTemplates = new Map<string, TemplateStringsArray>();
 
 /**
  * Interprets a template literal as an HTML template that can efficiently
  * render to and update a container.
  */
 export function html(strings: TemplateStringsArray, ...values: any[]): TemplateResult {
+  if (shouldCacheTemplates) {
+    const _key = strings.join('{{typescriptProblems}}');
```

Comment:
since the Template constructor already joins the literals parts, we could just reuse that as the key:

```ts
let key = cachesTemplates ? strings : strings.join(exprMarker);
let template = templates.get(key);
if (template === undefined) {
  template = new Template(strings, cachesTemplates ? undefined : key)
  templates.set(key, template);
}
```

```ts
class Template {
  constructor(strings: TemplateStringsArray, html?: string) { ... }
}
```

Reply:
`exprMarker` uses `Math.random()` which means the key will always be different. I also isolated the code with the assumption that it will be deleted when TypeScript follows the spec. We can also RegExp the `exprMarker` in the `Template` with one using `Math.random()`